File: ORIGINAL_4991/Phase III/FYP web 2/frontend/app/models/workflow.ts
// Create this file: frontend/models/workflow.ts

// These interfaces should correspond to the Pydantic models defined in backend/app/models/workflow.py

export interface WorkflowConfig {
  // Fields that the user can configure via the API
  strategy_name?: string | null;
  seed?: number | null;
  start_date_str?: string | null;           // Expects YYYY-MM-DD format string
  validation_end_date?: string | null;    // Expects YYYY-MM-DD format string
  extend_simulation_end_date_str?: string | null; // Expects YYYY-MM-DD format string
  lookback_date_str?: string | null;      // Expects YYYY-MM-DD format string
  tickers?: string[] | null;            // List of stock tickers
  k_values?: number[] | null;           // List of k values for grid search (maps to K_VALUES_CANDIDATES)
  n_values?: number[] | null;           // List of n values for grid search (maps to N_VALUES_CANDIDATES)
  total_strategy_capital?: number | null; // Starting capital
  // Add other optional overrides if defined in backend Pydantic model (e.g., epochs, learning_rate)
}

export interface WorkflowRunResponse {
  // Response when starting a workflow
  message: string;
  task_id: string;
  run_id: string; // Often the same as task_id
}

export interface WorkflowStatusResponse {
  // Response from the status endpoint
  task_id: string;
  run_id: string;
  status: 'PENDING' | 'STARTED' | 'PROGRESS' | 'SUCCESS' | 'FAILURE' | 'RETRY' | 'REVOKED' | string; // Allow other Celery states as string
  details?: {
    message?: string;
    step?: string;      // Current step description
    progress?: number;  // Overall progress percentage (0-100)
    error?: string;     // Error message string on failure
    traceback?: string; // Optional traceback string on failure
    best_k?: number;    // Might be available in details on success/progress
    best_n?: number;
    val_loss?: number;  // Example: Include validation loss during training progress
    [key: string]: any; // Allow other arbitrary details from Celery meta
  } | null;
}

export interface FileInfo {
  // Relative paths to generated files within the run's output directory
  metrics_csv?: string | null;
  snapshot_csv?: string | null;
  report_txt?: string | null;
  training_history_plot?: string | null;
  full_period_plot?: string | null;
  equity_curve_plot?: string | null;
  test_period_plot?: string | null;
  log_file?: string | null;
}

export interface WorkflowResultResponse {
  // Response from the results endpoint for a completed run
  run_id: string;
  status: 'SUCCESS' | 'FAILURE'; // Final status
  message: string;
  best_k?: number | null;
  best_n?: number | null;
  output_dir_url: string; // Base URL to access files for this specific run
  files: FileInfo;        // Dictionary containing relative file paths
  error_details?: string | null; // Full traceback or detailed error on failure
}

// --- History Related Models ---
export interface RunHistoryItem {
     // Structure for a single item in the run history list
     run_id: string;
     start_time?: string | null; // ISO Format Date string (from backend JSON)
     end_time?: string | null;   // ISO Format Date string (from backend JSON)
     status: string;            // Celery status string (e.g., SUCCESS, FAILURE)
     best_k?: number | null;
     best_n?: number | null;
     message?: string | null;    // Short summary or error message
}

export interface RunHistoryResponse {
     // Response from the history endpoint
     history: RunHistoryItem[];
}

// --- CSV Data Structure Models ---
export interface MetricData {
    // Structure for data parsed from the metrics CSV
    Metric: string;                    // Name of the metric
    Strategy: string | number | null;  // Value for the strategy
    // Allow dynamic keys for Benchmark and Tracker columns
    [key: string]: string | number | null;
}

export interface SnapshotData {
    // Structure for data parsed from the snapshot CSV
    Date: string;                      // Date of the snapshot (ISO format string)
    Category: string;                  // e.g., 'Top Gainer', 'Current Holding'
    Rank?: number | string | null;     // Rank for movers (can be string 'N/A' or number)
    StockID: string;                   // Ticker symbol
    'Cumulative Return'?: number | string | null; // Parsed value (can be string 'N/A')
    'Snapshot Date Price'?: number | string | null;
    'Snapshot Date Daily Return'?: number | string | null;
    'Snapshot Date Score'?: number | string | null;
}

File: ORIGINAL_4991/Phase III/FYP web 2/backend/app/models/workflow.py
# backend/app/models/workflow.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import datetime

class WorkflowConfig(BaseModel):
    # K_VALUES and N_VALUES REMOVED from here
    strategy_name: Optional[str] = Field(default=None, description="Custom name for this run")
    seed: Optional[int] = Field(default=None, description="Random seed for reproducibility")
    start_date_str: Optional[str] = Field(default=None, pattern=r'^\d{4}-\d{2}-\d{2}$', description="Start date (YYYY-MM-DD)")
    validation_end_date: Optional[str] = Field(default=None, pattern=r'^\d{4}-\d{2}-\d{2}$', description="End date for training/validation data (YYYY-MM-DD)")
    extend_simulation_end_date_str: Optional[str] = Field(default=None, pattern=r'^\d{4}-\d{2}-\d{2}$', description="End date for extended simulation (YYYY-MM-DD)")
    lookback_date_str: Optional[str] = Field(default=None, pattern=r'^\d{4}-\d{2}-\d{2}$', description="Date for snapshot analysis (YYYY-MM-DD)")
    tickers: Optional[List[str]] = Field(default=None, description="List of ticker symbols")
    # k_values: Optional[List[int]] = Field(default=None...) # REMOVED
    # n_values: Optional[List[int]] = Field(default=None...) # REMOVED
    total_strategy_capital: Optional[float] = Field(default=None, gt=0, description="Initial capital for simulation")

    model_config = { # Example remains the same, just won't include k/n
        "json_schema_extra": {
            "examples": [
                {
                    "strategy_name": "My HSI Run April (Defaults)",
                    "total_strategy_capital": 500000
                }
            ]
        }
    }

# ... rest of the models remain the same ...

class WorkflowRunResponse(BaseModel):
    message: str
    task_id: str
    run_id: str

class WorkflowStatusResponse(BaseModel):
    task_id: str
    run_id: str
    status: str
    details: Optional[Dict[str, Any]] = None

class FileInfo(BaseModel):
    metrics_csv: Optional[str] = None
    snapshot_csv: Optional[str] = None
    report_txt: Optional[str] = None
    training_history_plot: Optional[str] = None
    full_period_plot: Optional[str] = None
    equity_curve_plot: Optional[str] = None
    test_period_plot: Optional[str] = None
    log_file: Optional[str] = None

class WorkflowResultResponse(BaseModel):
    run_id: str
    status: str
    message: str
    best_k: Optional[int] = None # Still include best_k/n in RESULTS
    best_n: Optional[int] = None # Still include best_k/n in RESULTS
    output_dir_url: str
    files: FileInfo
    error_details: Optional[str] = None

class DefaultConfigResponse(BaseModel):
    # Ensure this returns relevant fields excluding k/n list inputs
    defaults: Dict[str, Any]

class RunHistoryItem(BaseModel):
     run_id: str
     start_time: Optional[datetime] = None
     end_time: Optional[datetime] = None
     status: str
     best_k: Optional[int] = None
     best_n: Optional[int] = None

class RunHistoryResponse(BaseModel):
     history: List[RunHistoryItem]

class RunHistoryItem(BaseModel):
     run_id: str
     start_time: Optional[datetime] = None # Often not easily available without DB
     end_time: Optional[datetime] = None
     status: str
     best_k: Optional[int] = None
     best_n: Optional[int] = None
     message: Optional[str] = None # Short summary or error message

class RunHistoryResponse(BaseModel):
     history: List[RunHistoryItem]
